src/Development/IDE/Core/Preprocessor.hs

```diff
@@ -0,0 +1,133 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Development.IDE.Core.Preprocessor
+  ( preprocessor
+  ) where
+
+import Development.IDE.GHC.CPP
+import Development.IDE.GHC.Orphans()
+import Development.IDE.GHC.Compat
+import GHC
+import GhcMonad
+import StringBuffer as SB
+
+import Data.List.Extra
+import System.FilePath
+import System.IO.Extra
+import Data.Char
+import DynFlags
+import qualified HeaderInfo as Hdr
+import Development.IDE.Types.Diagnostics
+import Development.IDE.GHC.Error
+import SysTools (Option (..), runUnlit)
+import Control.Monad.Trans.Except
+import qualified GHC.LanguageExtensions as LangExt
+import Data.Maybe
+
+
+-- | Given a file and some contents, apply any necessary preprocessors,
+--   e.g. unlit/cpp. Return the resulting buffer and the DynFlags it implies.
+preprocessor :: GhcMonad m => FilePath -> Maybe StringBuffer -> ExceptT [FileDiagnostic] m (StringBuffer, DynFlags)
+preprocessor filename mbContents = do
+    -- Perform unlit
+    (isOnDisk, contents) <- if isLiterate filename then do
```

Comment:
The indentation here confuses me, I would prefer if we put `then` on a new line, indent both `then` and `else` and align the two.